File: modules/device-resource-telemetry/android/src/main/java/cloud/tenka/passport/deviceresourcetelemetry/TenkaDeviceResourceTelemetryModule.kt
package cloud.tenka.passport.deviceresourcetelemetry

import android.app.ActivityManager
import android.content.Context
import android.os.BatteryManager
import android.os.Build
import android.os.Debug
import android.os.PowerManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlin.math.min

class TenkaDeviceResourceTelemetryModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("TenkaDeviceResourceTelemetry")

    AsyncFunction("getSnapshot") {
      val context = appContext.reactContext
        ?: return@AsyncFunction unavailableSnapshot()
      val activityManager =
        context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
      val memoryInfo = ActivityManager.MemoryInfo()
      activityManager.getMemoryInfo(memoryInfo)
      val residentMemoryBytes = Debug.getPss().toLong() * 1_024L
      // Issue 104 PR #132（Codex 指摘 major）: `memoryInfo.availMem` は端末全体の
      // 空き容量であり、この App 専用の割当上限ではない（他 App の状態次第で
      // 変動する。Android には iOS の `os_proc_available_memory()` に相当する
      // Process 単位の公開 API が無い）。TS 側（`local-model-manifest.ts` の
      // `evaluateModelResourceRisk`）へ `system-wide-available` として伝え、
      // iOS の `os-process-ceiling`（Process 単位の実測 Ceiling）と同じ信頼度
      // で扱わせない（保守的な割引はドメイン層の責務にし、Native 側の値自体は
      // 偽装しない）。
      val processCeiling = min(
        memoryInfo.totalMem,
        memoryInfo.availMem + residentMemoryBytes
      )

      mapOf(
        "physicalMemoryBytes" to positiveOrNull(memoryInfo.totalMem),
        "processMemoryLimitBytes" to positiveOrNull(processCeiling),
        "processMemoryLimitProvenance" to "system-wide-available",
        "processMemoryBytes" to positiveOrNull(residentMemoryBytes),
        "thermalState" to thermalState(context),
        "batteryLevelPermille" to batteryLevelPermille(context)
      )
    }
  }

  private fun unavailableSnapshot(): Map<String, Any?> = mapOf(
    "physicalMemoryBytes" to null,
    "processMemoryLimitBytes" to null,
    "processMemoryLimitProvenance" to "unavailable",
    "processMemoryBytes" to null,
    "thermalState" to "unknown",
    "batteryLevelPermille" to null
  )

  private fun positiveOrNull(value: Long): Long? = value.takeIf { it > 0L }

  private fun thermalState(context: Context): String {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) return "unknown"
    val powerManager =
      context.getSystemService(Context.POWER_SERVICE) as PowerManager
    return when (powerManager.currentThermalStatus) {
      PowerManager.THERMAL_STATUS_NONE -> "nominal"
      PowerManager.THERMAL_STATUS_LIGHT -> "fair"
      PowerManager.THERMAL_STATUS_MODERATE -> "fair"
      PowerManager.THERMAL_STATUS_SEVERE -> "serious"
      PowerManager.THERMAL_STATUS_CRITICAL -> "critical"
      PowerManager.THERMAL_STATUS_EMERGENCY -> "critical"
      PowerManager.THERMAL_STATUS_SHUTDOWN -> "critical"
      else -> "unknown"
    }
  }

  private fun batteryLevelPermille(context: Context): Int? {
    val batteryManager =
      context.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
    val percent = batteryManager.getIntProperty(
      BatteryManager.BATTERY_PROPERTY_CAPACITY
    )
    return percent.takeIf { it in 0..100 }?.times(10)
  }
}

File: modules/device-resource-telemetry/ios/TenkaDeviceResourceTelemetryModule.swift
import ExpoModulesCore
import MachO
import UIKit
import os

public final class TenkaDeviceResourceTelemetryModule: Module {
  public func definition() -> ModuleDefinition {
    Name("TenkaDeviceResourceTelemetry")

    AsyncFunction("getSnapshot") { () -> [String: Any?] in
      let footprint = self.physFootprintBytes()
      let physical = Int64(ProcessInfo.processInfo.physicalMemory)
      let device = UIDevice.current
      device.isBatteryMonitoringEnabled = true
      let batteryLevel = device.batteryLevel

      return [
        "physicalMemoryBytes": physical,
        // Issue 104 Priority 2（Bonsai-ready 化）: `os_proc_available_memory()`
        // （iOS の public C API、`<os/proc.h>`）で実測する。以前はここを常に
        // `nil` にしていた（推測値で偽装しないため）が、Bonsai 27B 級の大きな
        // Model を起動可否ゲート（`evaluateModelResourceRisk`）へ接続するには
        // 実測値が要る。code-reviewer 指摘: `os_proc_available_memory()` 単体は
        // 「あと確保できる残り」であり、Android 実装（`processCeiling = min(
        // totalMem, availMem + residentMemoryBytes)`）が返す「現在の使用量を
        // 含めた到達可能上限」とは意味が異なる（同じ effectiveMemoryBytes 計算
        // 式で分母として扱われるため、意味を揃える）。`phys_footprint`（現在の
        // 使用量）を足し戻し、`physicalMemoryBytes` を上限に丸めて Android と
        // 同じ「到達しうる Ceiling」を返す。
        "processMemoryLimitBytes": self.processMemoryCeilingBytes(
          footprint: footprint,
          physicalMemoryBytes: physical
        ),
        // Issue 104 PR #132（Codex 指摘 major）: この値は `os_proc_available_memory()`
        // （Process 単位の実測 Ceiling）由来であり、Android の `system-wide-available`
        // （端末全体の空き容量、Process 専用の割当上限ではない）とは信頼度が異なる。
        // TS 側（`local-model-manifest.ts` の `evaluateModelResourceRisk`）がこの
        // provenance を見て Android 値だけを保守的に割り引く。
        "processMemoryLimitProvenance": "os-process-ceiling",
        "processMemoryBytes": footprint,
        "thermalState": self.thermalState(),
        "batteryLevelPermille": batteryLevel < 0
          ? nil
          : Int((batteryLevel * 1_000).rounded())
      ]
    }.runOnQueue(.main)
  }

  /**
   * Issue 104 Priority 2（Bonsai-ready 化）: `task_info(MACH_TASK_BASIC_INFO)` の
   * `resident_size` は compressed memory を含まないため、Xcode の Memory Gauge・
   * Apple 自身が「実際のメモリ使用量」の指標として推奨する `phys_footprint`
   * （`task_info(TASK_VM_INFO)`）へ置き換える。
   */
  private func physFootprintBytes() -> Int64? {
    var info = task_vm_info_data_t()
    var count = mach_msg_type_number_t(
      MemoryLayout<task_vm_info_data_t>.size / MemoryLayout<integer_t>.size
    )
    let result = withUnsafeMutablePointer(to: &info) { pointer in
      pointer.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
        task_info(
          mach_task_self_,
          task_flavor_t(TASK_VM_INFO),
          $0,
          &count
        )
      }
    }
    guard result == KERN_SUCCESS, count >= TASK_VM_INFO_REV1_COUNT else {
      return nil
    }
    return Int64(info.phys_footprint)
  }

  /**
   * この Process が OOM Kill されるまでにあと確保できる概算 byte 数
   * （iOS の public C API、`os_proc_available_memory()`）。呼び出しごとの
   * 瞬間値であり、キャッシュしない（Apple のドキュメント上の注意事項）。
   */
  private func availableMemoryBytes() -> Int64? {
    let available = os_proc_available_memory()
    return available > 0 ? Int64(available) : nil
  }

  /**
   * 「現在の使用量（`phys_footprint`）+ あと確保できる残り
   * （`os_proc_available_memory()`）」を、Android 実装
   * （`TenkaDeviceResourceTelemetryModule.kt` の `processCeiling`）と同じ
   * 「この Process が到達しうる Memory の上限」という意味に揃える。
   * `physicalMemoryBytes` を超えないよう丸める（Android の `min(totalMem, ...)`
   * と同じ安全策）。どちらかを実測できない場合は `nil`（推測値で埋めない）。
   */
  private func processMemoryCeilingBytes(
    footprint: Int64?,
    physicalMemoryBytes: Int64
  ) -> Int64? {
    guard let footprint = footprint,
          let available = availableMemoryBytes() else {
      return nil
    }
    return min(physicalMemoryBytes, footprint + available)
  }

  private func thermalState() -> String {
    switch ProcessInfo.processInfo.thermalState {
    case .nominal:
      return "nominal"
    case .fair:
      return "fair"
    case .serious:
      return "serious"
    case .critical:
      return "critical"
    @unknown default:
      return "unknown"
    }
  }
}
